### PR TITLE
[new release] crlibm (0.5)

### PR DESCRIPTION
**Pull request by Chris00**
Binding to CRlibm, a correctly rounded math lib

- Project page: <a href="https://github.com/Chris00/ocaml-crlibm">https://github.com/Chris00/ocaml-crlibm</a>
- Documentation: <a href="https://Chris00.github.io/ocaml-crlibm/doc">https://Chris00.github.io/ocaml-crlibm/doc</a>

##### CHANGES:

- Rename sub-modules to more standard names.

### CODE REVIEW

**Comment by Chris00** [comment]
Not sure what the problem is — it builds fine on my machine and by [Github Actions](https://github.com/Chris00/ocaml-crlibm/actions/runs/1418759038).

**Comment by mseri** [comment]
Looks like some kind of corruption  in the Ci? Can it be?

**Comment by kit-ty-kate** [comment]
The package failed to build on x86_32:
```
#=== ERROR while compiling crlibm.0.5 =========================================#
# context              2.1.0 | linux/x86_32 | ocaml-base-compiler.4.13.1 | pinned(https://github.com/Chris00/ocaml-crlibm/releases/download/0.5/crlibm-0.5.tbz)
# path                 ~/.opam/4.13/.opam-switch/build/crlibm.0.5
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p crlibm -j 31 @runtest
# exit-code            1
# env-file             ~/.opam/log/crlibm-77-28f843.env
# output-file          ~/.opam/log/crlibm-77-28f843.out
### output ###
#     ocamlopt tests/test.exe (exit 2)
# (cd _build/default && /home/opam/.opam/4.13/bin/ocamlopt.opt -w -40 -g -o tests/test.exe src/crlibm.cmxa -I src /home/opam/.opam/4.13/lib/ocaml/unix.cmxa -I /home/opam/.opam/4.13/lib/ocaml /home/opam/.opam/4.13/lib/benchmark/benchmark.cmxa tests/.test.eobjs/native/test.cmx)
# /usr/bin/ld: tests/.test.eobjs/native/test.o: warning: relocation in read-only section `.text'
# /usr/bin/ld: src/crlibm.a(crlibm.o): in function `camlCrlibm__fun_346':
# :(.text+0x5c): undefined reference to `log10_rn'
# /usr/bin/ld: src/crlibm.a(crlibm.o): in function `camlCrlibm__fun_348':
# :(.text+0xac): undefined reference to `log2_rn'
# /usr/bin/ld: src/crlibm.a(crlibm.o): in function `camlCrlibm__fun_426':
# /home/opam/.opam/4.13/.opam-switch/build/crlibm.0.5/_build/default/src/crlibm.ml:67: undefined reference to `log10_rd'
# /usr/bin/ld: src/crlibm.a(crlibm.o): in function `camlCrlibm__fun_424':
# /home/opam/.opam/4.13/.opam-switch/build/crlibm.0.5/_build/default/src/crlibm.ml:67: undefined reference to `log2_rd'
# /usr/bin/ld: src/crlibm.a(crlibm.o): in function `camlCrlibm__fun_466':
# /home/opam/.opam/4.13/.opam-switch/build/crlibm.0.5/_build/default/src/crlibm.ml:107: undefined reference to `log10_ru'
# /usr/bin/ld: src/crlibm.a(crlibm.o): in function `camlCrlibm__fun_464':
# /home/opam/.opam/4.13/.opam-switch/build/crlibm.0.5/_build/default/src/crlibm.ml:107: undefined reference to `log2_ru'
# /usr/bin/ld: src/crlibm.a(crlibm.o): in function `camlCrlibm__fun_506':
# /home/opam/.opam/4.13/.opam-switch/build/crlibm.0.5/_build/default/src/crlibm.ml:147: undefined reference to `log10_rz'
# /usr/bin/ld: src/crlibm.a(crlibm.o): in function `camlCrlibm__fun_504':
# /home/opam/.opam/4.13/.opam-switch/build/crlibm.0.5/_build/default/src/crlibm.ml:147: undefined reference to `log2_rz'
# /usr/bin/ld: src/crlibm.a(crlibm.o): in function `camlCrlibm__1':
# :(.data+0x464): undefined reference to `log10_rz'
# /usr/bin/ld: :(.data+0x468): undefined reference to `log2_rz'
# /usr/bin/ld: :(.data+0x4b4): undefined reference to `log10_ru'
# /usr/bin/ld: :(.data+0x4b8): undefined reference to `log2_ru'
# /usr/bin/ld: :(.data+0x504): undefined reference to `log10_rd'
# /usr/bin/ld: :(.data+0x508): undefined reference to `log2_rd'
# /usr/bin/ld: :(.data+0x558): undefined reference to `log10_rn'
# /usr/bin/ld: :(.data+0x55c): undefined reference to `log2_rn'
# /usr/bin/ld: src/libcrlibm_stubs.a(crlibm_stubs.o): in function `log2_rn_bc':
# /home/opam/.opam/4.13/.opam-switch/build/crlibm.0.5/_build/default/src/crlibm_stubs.c:49: undefined reference to `log2_rn'
# /usr/bin/ld: src/libcrlibm_stubs.a(crlibm_stubs.o): in function `log2_rd_bc':
# /home/opam/.opam/4.13/.opam-switch/build/crlibm.0.5/_build/default/src/crlibm_stubs.c:49: undefined reference to `log2_rd'
# /usr/bin/ld: src/libcrlibm_stubs.a(crlibm_stubs.o): in function `log2_ru_bc':
# /home/opam/.opam/4.13/.opam-switch/build/crlibm.0.5/_build/default/src/crlibm_stubs.c:49: undefined reference to `log2_ru'
# /usr/bin/ld: src/libcrlibm_stubs.a(crlibm_stubs.o): in function `log2_rz_bc':
# /home/opam/.opam/4.13/.opam-switch/build/crlibm.0.5/_build/default/src/crlibm_stubs.c:49: undefined reference to `log2_rz'
# /usr/bin/ld: src/libcrlibm_stubs.a(crlibm_stubs.o): in function `log10_rn_bc':
# /home/opam/.opam/4.13/.opam-switch/build/crlibm.0.5/_build/default/src/crlibm_stubs.c:50: undefined reference to `log10_rn'
# /usr/bin/ld: src/libcrlibm_stubs.a(crlibm_stubs.o): in function `log10_rd_bc':
# /home/opam/.opam/4.13/.opam-switch/build/crlibm.0.5/_build/default/src/crlibm_stubs.c:50: undefined reference to `log10_rd'
# /usr/bin/ld: src/libcrlibm_stubs.a(crlibm_stubs.o): in function `log10_ru_bc':
# /home/opam/.opam/4.13/.opam-switch/build/crlibm.0.5/_build/default/src/crlibm_stubs.c:50: undefined reference to `log10_ru'
# /usr/bin/ld: src/libcrlibm_stubs.a(crlibm_stubs.o): in function `log10_rz_bc':
# /home/opam/.opam/4.13/.opam-switch/build/crlibm.0.5/_build/default/src/crlibm_stubs.c:50: undefined reference to `log10_rz'
# /usr/bin/ld: warning: creating DT_TEXTREL in a PIE
# collect2: error: ld returned 1 exit status
# File "caml_startup", line 1:
# Error: Error during linking (exit code 1)
```
Is that expected?

**Comment by mseri** [comment]
Big differences in output of successful and failing builds are the ld warning
```
- /usr/bin/ld: src/crlibm.a(crlibm.o): warning: relocation in read-only section `.text'
- /usr/bin/ld: warning: creating DT_TEXTREL in a shared object
```
and the fact that the failing build is not complaining about `/0` for all the log2 and log10 functions

**Comment by mseri** [comment]
Let me check if there is a CI issue, I  am seeing similar failures on other packages as well

**Comment by Chris00** [comment]
> Big differences in output of successful and failing builds are the ld warning (present only on the 32bit build)

I saw that but I do not get it: [everything is compiled with `-fPIC`](https://gcc-help.gcc.gnu.narkive.com/MSci4Gkx/dt-textrel-how-to-avoid-this-warning).

**Comment by mseri** [comment]
Yes, I saw, that part of the reason why I start believing that these failures are a CI issue. I am seeing similar ones also in other packages

**Comment by mooreryan** [comment]
PIC issues may be related to this: https://github.com/ocaml/ocaml/issues/9800

**Comment by Chris00** [comment]
@mseri I ran the following docker file
```
FROM ocaml/opam:debian-11-ocaml-4.13
SHELL [ "/usr/bin/linux32", "/bin/sh", "-c" ]
USER 1000:1000
RUN opam repository remove -a multicore || true
RUN sudo ln -f /usr/bin/opam-2.1 /usr/bin/opam
RUN opam init --reinit -ni
ENV OPAMDOWNLOADJOBS 1
ENV OPAMERRLOGLEN 0
ENV OPAMSOLVERTIMEOUT 500
ENV OPAMPRECISETRACKING 1
WORKDIR /home/opam
RUN rm -rf opam-repository/
COPY --chown=1000:1000 . crlibm/
RUN opam pin add -k path -yn crlibm.dev crlibm/
RUN opam update --depexts
RUN opam install --deps-only --with-test crlibm && opam install -v --with-test crlibm
```
but my machine is 64 bits and it runs fine.

**Comment by mseri** [comment]
That should run in 32 bits even on a 64-bit machine, or maybe you need to build it with
```
docker build --platform=linux/386 -f Dockerfile .
```

i will also try asap

**Comment by Chris00** [comment]
This yields:
```
failed to get destination image "sha256:871a1e5575f19b25e99f83b82c1a3ed24b14d295120f5937dc79586d13846bec": image with reference sha256:871a1e5575f19b25e99f83b82c1a3ed24b14d295120f5937dc79586d13846bec was found but does not match the specified platform: wanted linux/386, actual: linux/amd64
```

File: packages/crlibm/crlibm.0.5/opam [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  ["dune" "build" "-p" name "-j" jobs "@runtest"] {with-test}
+]
+depends: [
+  "ocaml" {>= "4.03"}
+  "dune-configurator" {>= "2.0"}
+  "dune" {>= "2.0"}
+  "base-bytes" {build}
+  "benchmark" {with-test}
+]
+synopsis: "Binding to CRlibm, a correctly rounded math lib"
```
Comment (mseri) [inline packages/crlibm/crlibm.0.5/opam:22]:
```suggestion
synopsis: "Binding to CRlibm, a correctly rounded math lib"
available: arch != "x86_32"
```

**Comment by mseri** [comment]
The first command to run the package was correct, without the platform.
To reproduce our failures you can run
```
git clone --recursive "https://github.com/ocaml/opam-repository.git" && cd "opam-repository" && git fetch origin "refs/pull/19943/head" && git reset --hard e553d945
git fetch origin master
git merge 87ef72b5cd492573258eb1b6f3b30c88af31ae3f
cat > ../Dockerfile <<'END-OF-DOCKERFILE'
FROM ocaml/opam:debian-11-ocaml-4.13@sha256:0754672dc08f90477fff95b6489b92ff6020e63304fd1eabae57147ec40c4e90
SHELL [ "/usr/bin/linux32", "/bin/sh", "-c" ]
USER 1000:1000
RUN opam repository remove -a multicore || true
RUN sudo ln -f /usr/bin/opam-2.1 /usr/bin/opam
RUN opam init --reinit -ni
ENV OPAMDOWNLOADJOBS 1
ENV OPAMERRLOGLEN 0
ENV OPAMSOLVERTIMEOUT 500
ENV OPAMPRECISETRACKING 1
WORKDIR /home/opam
RUN rm -rf opam-repository/
COPY --chown=1000:1000 . opam-repository/
RUN opam repository set-url --strict default "file://$HOME/opam-repository"
RUN opam pin add -k version -yn crlibm.0.5 0.5
RUN opam update --depexts
RUN opam remove crlibm.0.5 && opam install --deps-only crlibm.0.5 && opam install -v crlibm.0.5; \
    res=$?; \
    test "$res" != 31 && exit "$res"; \
    export OPAMCLI=2.0; \
    build_dir=$(opam var prefix)/.opam-switch/build; \
    failed=$(ls "$build_dir"); \
    for pkg in $failed; do \
    if opam show -f x-ci-accept-failures: "$pkg" | grep -qF "\"debian-11\""; then \
    echo "A package failed and has been disabled for CI using the 'x-ci-accept-failures' field."; \
    fi; \
    done; \
    exit 1
RUN opam update --depexts
RUN opam remove crlibm.0.5 && opam install --deps-only --with-test crlibm.0.5 && opam install -v --with-test crlibm.0.5; \
    res=$?; \
    test "$res" != 31 && exit "$res"; \
    export OPAMCLI=2.0; \
    build_dir=$(opam var prefix)/.opam-switch/build; \
    failed=$(ls "$build_dir"); \
    for pkg in $failed; do \
    if opam show -f x-ci-accept-failures: "$pkg" | grep -qF "\"debian-11\""; then \
    echo "A package failed and has been disabled for CI using the 'x-ci-accept-failures' field."; \
    fi; \
    done; \
    exit 1

END-OF-DOCKERFILE
docker build -f ../Dockerfile .
```

You can execute the images that this produce to test your code with
```
docker run -it last_successfully_built_step_hash_in_the_output /usr/bin/linux32 bash
```

**Comment by kit-ty-kate** [comment]
Thanks! Sorry for the wait

**Comment by Chris00** [comment]
Thanks for your help!

**Pull request merged**